docs/code/CrossPlatform.md

```diff
@@ -0,0 +1,179 @@
+ML.NET has a hard dependency on x86/x64. Some of the dependency is on Intel MKL, while other parts depend on x86/x64 SIMD instructions. To make things easier I will refer to these as just the x86/x64 dependencies. This is to perform many optimized math functions and enables several transformers/trainers to be run natively for improved performance. The problem is that these dependencies can only run on x86/x64 machines and are the main blockers for expanding to other architectures. While you can run the managed code on other architectures, there is no good way to know which parts will run and which ones wont. This includes the build process as well, which currently has these same hard dependencies and building on non x86/x64 machines is not supported.
+
+ML.NET also has dependencies on things that either don't build on other architectures or have to be compiled by the user if its wanted. For example:
+ - LightGBM
+ - TensorFlow
+ - OnnxRuntime
+
+I will go over these in more depth below.
+
+### 2.2 Build
+Since ML.NET has a hard dependency on x86/x64, the build process assumes it's running there. For example, the build process will try and copy native dlls without checking if they exist because it assumes the build for them succeeded or that they are available. The build process will need to be modified so that it doesn't fail when it can't find these files. It does the same copy for our own Native dlls, so this will need to be fixed for those as well.
+
+### 2.3 Managed Code
+Since ML.NET has a hard dependency on x86/x64, the managed code imports dlls without checking whether or not they exist. If the dlls don't exist you get a hard failure. For example, if certain columns are active the `MulticlassClassificationScorer` will call `CalculateIntermediateVariablesNative` which is loaded from `CpuMathNative`, but all of this is done without any checks to see if the DLL actually exists. The tests also run into this probleb. The base test class imports and sets up Intel MKL even if the test itself does not need it.
+
+### 2.4 Native Projects
+ML.NET has 6 native projects. They are:
+ - CpuMathNative
+   - Partial managed fallback when using NetCore 3.1.
+   - A large amount of work would be required to port the native code to other platforms. We would have to change all the SIMD instructions for each platform.
+   - A small amount of work required for a full managed fallback.
+   - This was created before we had hardware intrinsics so we used native code for performance. The managed fallback uses x86/x64 intrinsics where possible.
+ - FastTreeNative
+   - Full managed fallback by changing a C# compiler flag. This flag is hardcoded to always use the native code.
+   - Small amount of work required to change build process and verify it's correct.
+   - This was created before we had hardware intrinsics so we used native code for performance. The managed fallback does not use hardware intrinsics, so this will be slower than the native solution.
```

Comment:
> The managed fallback does not use hardware intrinsics

But could with work, presumably?

Reply:
Yeah. Let me change the wording to say that the current state doesn't use them but they could be added in.